dataflow_test.py

```diff
@@ -114,23 +114,23 @@ def test_longer_pipeline():
 
 def test_fork_implicit_pipes():
 
-    # Arguments can be pipes or tuples
-    # tuples get implicitely converted into pipes
+    # Arguments can be pipes or tuples.
+    # Tuples get implicitely converted into pipes
 
     the_source = list(range(10, 20))
     add_1      = df.map(lambda x: 1 + x)
 
-    tuple_collector = []; tuple_sink = df.sink(tuple_collector.append)
-    pipe_collector  = [];  pipe_sink = df.sink( pipe_collector.append)
+    implicit_pipe_collector = []; implicit_pipe_sink = df.sink(implicit_pipe_collector.append)
+    explicit_pipe_collector = []; explicit_pipe_sink = df.sink(explicit_pipe_collector.append)
 
-    pipe_as_tuple =        (add_1, tuple_sink)
-    pipe_as_pipe  = df.pipe(add_1,  pipe_sink)
+    implicit_pipe =        (add_1, implicit_pipe_sink)
+    explicit_pipe = df.pipe(add_1, explicit_pipe_sink)
 
     df.push(source = the_source,
-            pipe   = df.fork(pipe_as_tuple,
-                             pipe_as_pipe ))
+            pipe   = df.fork(implicit_pipe,
+                             explicit_pipe))
```

Comment:
I think that implicit pipes are designed to reduce noise and number of characters typed in more complex graphs. So I don't think that you would ever write something like line 126 in real life, and I don't think we should encourage it over line 127. So, while I like the clarity of this test as you've written it, as a user-facing documentation test, I think I would prefer to combine 126-131 into

````python
df.push(source = the_source,
        pipe   = df.fork(       (add_1, implicit_pipe_sink),
                         df.pipe(add_1, explicit_pipe_sink))
````

Looking at it, I think that the sink names in there, get the point across quite well. What do you think?

Also, off the top of my head, I don't remember whether `push(pipe = ...)` accepts implicit pipes, and whether it's tested or not. If it doesn't it probably should.

Reply:
>  think I would prefer to combine 126-131 into
> ``` df.push(source = the_source,
>        pipe   = df.fork(       (add_1, implicit_pipe_sink),
>                         df.pipe(add_1, explicit_pipe_sink))
> ```
> Looking at it, I think that the sink names in there, get the point
> across quite well. What do you think?

I completely agree. It looks much better like that.

> Also, off the top of my head, I don't remember whether `push(pipe = ...)` accepts implicit pipes, and whether it's tested or not. If it doesn't it probably should.

I don't remember either. Will check and test if necessary.

Reply:
> I don't remember whether push(pipe = ...) accepts implicit pipes

It doesn't. Do you want it to? If so, do you want to include it in this PR?

Reply:
> Do you want it to?

I think that the `pipe` keyword argument makes it abundantly clear that it's a pipe, so the `df.pipe` that immediately follows it is just boilerplate/noise and needlessly increases the indentation of the graph by 7 or 4 characters. So I'd say yes.

I trust that the pipe argument is keyword-only ... hmm, it isn't. Do you agree that it (well, all three of `pipe`s admissible arguments) should be keyword-only? Hmm, maybe not. Need to ruminate on this.

> If so, do you want to include it in this PR?

No. Let's try to keep them small and orthogonal. This is already turning into a monster.

Reply:
> It is just boilerplate/noise and needlessly increases the indentation of the graph by 7 or 4 characters

I agree.

> Do you agree that it (well, all three of `push`'s admissible arguments) should be keyword-only?

I don't see any reason why not. It encourages (actually forces) a verbose and obvious-to-the-eye syntax. I am quite comfortable with making it keyword-only.

> No. Let's try to keep them small and orthogonal.

Good. @mmkekic told me she wanted to contribute. Maybe this is a good introduction point.

Reply:
OK, executive decision: `pipe`'s arguments should be keyword-only. Not in this PR.

>  @mmkekic told me she wanted to contribute.

Great!

> Maybe this is a good introduction point.

You mean get her to make these args keyword-only? It's literally adding 3 characters (including a space), and I'm confident that she's familiar with the style in which we'll be running this project, so it will be trivial for her and not especially enlightening ... but I have no objections. My only reservation is that I don't want to accept PRs until I'm happy that I won't be rebasing master any more ...

Reply:
> You mean get her to make these args keyword-only? 

No, I meant adding the implicit-pipe feature to the pipe argument. It's not a big deal either, but it will require a test and it's a bit more elaborate.

Reply:
Yup, excellent.